File: leetcode/array/904.fruit-into-baskets.py
class Solution:
    def totalFruit(self, tree: List[int]) -> int:
        if not tree:
            return []
        if len(tree) <= 2:
            return len(tree)

        i = 0
        j = 1
        result = [tree[i]]
        maxValue = 0
        while j < len(tree):
            if tree[j] in result:
                result.append(tree[j])
                j += 1
            elif tree[j] not in result and len(set(result)) != 2:
                result.append(tree[j])
                j += 1
            elif tree[j] not in result and len(set(result)) == 2:
                if maxValue < len(result):
                    maxValue = len(result)
                i = j-1
                while tree[i-1] == tree[i]:
                    i -= 1
                j = i+1
                result = [tree[i]]
            else:
                j += 1
        if len(result) > maxValue:
            maxValue = len(result)
        return maxValue

File: leetcodeJS/array/904.fruit-into-baskets.js
/**
 * @param {number[]} tree
 * @return {number}
 */
var totalFruit = function (tree) {
  if (tree.length == 0) return 0;
  if (tree.length < 3) return tree.length;
  let length = 0;
  let p = 0;
  let p2 = 1;
  let cnt = 1;
  let basket = [tree[p]];

  while (p2 < tree.length) {
    if (basket.includes(tree[p2])) basket.push(tree[p2]);
    else if (!basket.includes(tree[p2]) && cnt != 2) {
      cnt++;
      basket.push(tree[p2]);
    }
    if (!basket.includes(tree[p2]) && cnt == 2) {
      if (basket.length > length) length = basket.length;
      p = p2 - 1;
      while (tree[p] == tree[p - 1]) p--;
      p2 = p + 1;
      basket = [tree[p]];
      cnt = 1;
    } else {
      if (basket.length > length) length = basket.length;
      p2++;
    }
  }
  return length;
};
